### PR TITLE
Added contract types and refactored contract names in UI code

### LINKED ISSUE #354: Refactor UI to Interface With Updated Contract Method Names
Requires https://github.com/ubiquity/ubiquity-dollar/pull/351 to be merged in before this can be released to the DevPool.

> Frontend should be completely broken after these changes (need a task to fix it)

### PR DESCRIPTION
**Pull request by acaldas**
Added `typechain` to the build process to build the contract types.

Updated the UI code to enforce the contract types.

Resolves #354

### CODE REVIEW

**Comment by acaldas** [comment]
I'm not able to test since the deployed contracts are not up to date. Is there a script to deploy the contracts locally?

File: packages/dapp/components/redeem/UcrRedeem.tsx [hunk 1]
```diff
@@ -39,12 +39,12 @@ const UcrRedeem = ({ twapInteger }: { twapInteger: number }) => {
...
     const { debtCouponManager } = deployedContracts;
-    await ensureERC20Allowance("uCR -> DebtCouponManager", managedContracts.creditToken as unknown as Contract, amount, signer, debtCouponManager.address);
-    await (await debtCouponManager.connect(signer).burnAutoRedeemTokensForDollars(amount)).wait();
+    await ensureERC20Allowance("uCR -> DebtCouponManager", managedContracts.creditToken, amount, signer, debtCouponManager.address);
+    await (await debtCouponManager.connect(signer).burnCreditTokensForDollars(amount)).wait();
     refreshBalances();
     if (provider && quoteAmount && selectedRedeemToken !== "uAD") {
       const routerContract = getUniswapV3RouterContract(V3_ROUTER_ADDRESS, provider);
-      await (await routerContract.connect(signer).approveMax(quoteAmount, managedContracts.dollarToken)).wait();
+      await (await routerContract.connect(signer).approveMax(managedContracts.dollarToken.address)).wait();
```
Comment (acaldas) [inline packages/dapp/components/redeem/UcrRedeem.tsx:47]: `approveMax` was being called incorrectly.

**Comment by 0x4007** [comment]
> I'm not able to test since the deployed contracts are not up to date. Is there a script to deploy the contracts locally?

- I'm pretty sure that all of the deploy scripts are up-to-date. 
- I also believe that the deploys run as part of our CI. See https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/install-and-build-pr.yml

> [Deployment: Fri Dec 09 2022 17:39:09 GMT+0000 (Coordinated Universal Time)](https://639372b680601d00823613a7--ubiquity-dollar.netlify.app/)

- Lastly, the ubiquibot handles the deploy for you. Just click on the link to see that it doesn't work.

File: packages/dapp/pages/404.tsx [hunk 2]
```diff
@@ -9,22 +9,34 @@ const Custom404 = () => {
         </p>
         <ul>
           <li>
-            <a href="//github.com/ubiquity/ubiquity-dollar/wiki" target="_blank">Docs</a>
+            <a href="//github.com/ubiquity/ubiquity-dollar/wiki" target="_blank">
+              Docs
+            </a>
```
Comment (0x4007) [inline packages/dapp/pages/404.tsx:14]: I've seen this formatting being reverted several times. We must have a misconfig with one and/or another formatter. Can you confirm that this was done automatically by the formatter settings? If so, I will create a new issue to fix this.

Comment (acaldas) [reply packages/dapp/pages/404.tsx:14]: Yes, this came up as a lint warning, saving the file changes to this format.

**Comment by acaldas** [comment]
> > I'm not able to test since the deployed contracts are not up to date. Is there a script to deploy the contracts locally?
> 
> * I'm pretty sure that all of the deploy scripts are up-to-date.
> * I also believe that the deploys run as part of our CI. See https://github.com/ubiquity/ubiquity-dollar/blob/development/.github/workflows/install-and-build-pr.yml
> 
> > [Deployment: Fri Dec 09 2022 17:39:09 GMT+0000 (Coordinated Universal Time)](https://639372b680601d00823613a7--ubiquity-dollar.netlify.app/)
> 
> * Lastly, the ubiquibot handles the deploy for you. Just click on the link to see that it doesn't work.

I'm not very familiar with Foundry but I don't see where the contracts are deployed. Shouldn't `yarn workspace @ubiquity/contracts deploy ...` have to be run somewhere?

The deploy scripts refer to the old contracts, for instance: https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/scripts/deploy/dollar/UbiquityAlgorithmicDollarManager.ts#L13

**Comment by 0x4007** [comment]
> I'm not very familiar with Foundry but I don't see where the contracts are deployed. Shouldn't `yarn workspace @ubiquity/contracts deploy ...` have to be run somewhere?
> 
> The deploy scripts refer to the old contracts, for instance: https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/scripts/deploy/dollar/UbiquityAlgorithmicDollarManager.ts#L13

https://github.com/ubiquity/ubiquity-dollar/pull/367#issuecomment-1345492281

**Comment by acaldas** [comment]
> > I'm not very familiar with Foundry but I don't see where the contracts are deployed. Shouldn't `yarn workspace @ubiquity/contracts deploy ...` have to be run somewhere?
> > The deploy scripts refer to the old contracts, for instance: https://github.com/ubiquity/ubiquity-dollar/blob/development/packages/contracts/scripts/deploy/dollar/UbiquityAlgorithmicDollarManager.ts#L13
> 
> [#367 (comment)](https://github.com/ubiquity/ubiquity-dollar/pull/367#issuecomment-1345492281)

I was also forking mainnet with Hardhat to test locally but in this case, the contracts on mainnet are not updated so forking is not enough.
I see from that [comment ](https://github.com/ubiquity/ubiquity-dollar/pull/367#issuecomment-1345492281) that the scripts to deploy are being worked on so I should be able to test this after those changes.

**Comment by 0x4007** [comment]
Thanks for your patience sorry for the mess this week last week!

**Comment by 0x4007** [comment]
@zgorizzo69 would you mind updating @acaldas when he is unblocked here with your deployment scripts being ready? Thanks!

**Comment by zgorizzo69** [comment]
> @zgorizzo69 would you mind updating @acaldas when he is unblocked here with your deployment scripts being ready? Thanks!

yes I will try to push as much as I can on the diamond branch when I have code that actually compiles.
@acaldas  In the code (feat/diamond branch) you will find the deployment contract used to run the tests

**Comment by 0x4007** [comment]
Hey @zgorizzo69 requesting an update on this for @acaldas

**Pull request merged**